bitcoin/src/pow.rs

```diff
@@ -913,6 +929,57 @@ impl kani::Arbitrary for U256 {
     }
 }
 
+// FIXME: we should move this in a better place
+/// TODO: add some docs here
+#[cfg_attr(all(test, mutate), mutate)]
+pub fn calculate_next_work_required<F>(
+    last_height: Height,
+    last_time: u32,
+    last_target: Target,
+    adjustment_interval: u64,
+    pow_target_timespan: u64,
+    no_pow_retargeting: bool,
+    get_block_by_heigh: F,
+) -> u32
```

Comment:
In 003673d1cb0495b80fcea5e01333eba9594d8e20:

The closure can be changed to return a `BlockHeader` rather than a `Block`, which is much easier to deal with. Then if you provide `next_height` as a parameter you can collapse all the rest of the parameters:

* `last_height` is `next_height` minus one, doesn't need to be passed in
* `last_time` and `last_target` are in the blockheader for `last_height`, doesn't need to be passed in
* `adjustment_interval`, `pow_target_timespan`, `no_pow_retargeting` I think can all be set by a single`Network` argument to select between testnet/bitcoin/signet.

Next `get_block_by_height` needs to return a generic `Result` which should be propagated. Otherwise the user has no way to handle the case that they are missing blocks, requesting an invalid height, etc. If the user is sure their function can't fail they can provide `std::convert::Infallible` as the error type.

Reply:
Also, you have a typo `get_block_by_heigh` :).

Reply:
Also -- in the testnet case there is some rule where if there is no block for a long time the difficulty drops to 1, or something. It's fine to leave this as a TODO, but if we want to make this function generic then we should at least add a TODO comment.

Alternately, I'd be fine if we simply only supported Bitcoin ... then we could drop `adjustment_interval`, `pow_target_timespan` and `no_pow_retargeting` and just use fixed constants.

Reply:
Ok here is what I found after applying your suggestion

- where is the `next_height`? I am not sure I am using this as a procedure parameter, do I?
- The callback is already returning a Block Header, I just added the wrong name into it, or I am missing something?
- to remove `adjustment_interval`, `pow_target_timespan`, `no_pow_retargeting` I can use the `Params` struct?

>Also -- in the testnet case there is some rule where if there is no block for a long time the difficulty drops to 1, or something. It's fine to leave this as a TODO, but if we want to make this function generic then we should at least add a TODO comment.

If I want to support this in this PR what is the best approach, we return 1 if the network is testnet and there is no block generated for a long time?

Reply:
Oh, yeah, using the `Param`s struct is a good idea. And feel free to add fields to it if you need.

Reply:
As for `next_height`, yeah, oops, I'm not sure what I was thinking.

I would suggest renaming `last_height` to `current_height` though.

Reply:
Done!